DirectXTex/DirectXTexConvert.cpp

```diff
@@ -2627,6 +2627,178 @@ HRESULT DirectX::_ConvertFromR32G32B32A32(
 }
 
 
+//-------------------------------------------------------------------------------------
+// Convert DXGI image to/from GUID_WICPixelFormat64bppRGBAHalf (no range conversions)
+//-------------------------------------------------------------------------------------
+_Use_decl_annotations_
+HRESULT DirectX::_ConvertToR16G16B16A16(const Image& srcImage, ScratchImage& image)
+{
+    if (!srcImage.pixels)
+        return E_POINTER;
+
+    HRESULT hr = image.Initialize2D(DXGI_FORMAT_R16G16B16A16_FLOAT, srcImage.width, srcImage.height, 1, 1);
+    if (FAILED(hr))
+        return hr;
+
+    ScopedAlignedArrayXMVECTOR scanline(static_cast<XMVECTOR*>(_aligned_malloc((sizeof(XMVECTOR) * srcImage.width), 16)));
+    if (!scanline)
+    {
+        image.Release();
+        return E_OUTOFMEMORY;
+    }
+
+    const Image *img = image.GetImage(0, 0, 0);
+    if (!img)
+    {
+        image.Release();
+        return E_POINTER;
+    }
+
+    uint8_t* pDest = img->pixels;
+    if (!pDest)
+    {
+        image.Release();
+        return E_POINTER;
+    }
+
+    const uint8_t *pSrc = srcImage.pixels;
+    for (size_t h = 0; h < srcImage.height; ++h)
+    {
+        if (!_LoadScanline(scanline.get(), srcImage.width, pSrc, srcImage.rowPitch, srcImage.format))
+        {
+            image.Release();
+            return E_FAIL;
+        }
+
+        if (!_StoreScanline(pDest, img->rowPitch, DXGI_FORMAT_R16G16B16A16_FLOAT, scanline.get(), srcImage.width))
```

Comment:
I could use DirectXMath's ``XMConvertFloatToHalfStream`` here, but I want to actually try out some test cases first. The main reason to use the ``_StoreScanline`` solution is that it explicitly clamps the values into the float16 min/max range, while the DirectXMath conversion saturates to infinity.

Reply:
Problem cases are very large images in the following formats (non-WIC): ``DXGI_FORMAT_R32G32B32A32_UINT``, ``DXGI_FORMAT_R32G32B32A32_SINT``, ``DXGI_FORMAT_R32G32B32_UINT``, and ``DXGI_FORMAT_R32G32B32_SINT`` which exceed the range of a float16.